### PR TITLE
[REF/IMP] pivot table functions

### PR DESCRIPTION
**Pull request by hokolomopo**
## [REF] spreadsheet: define function in constants

Ths commit change the pivot/list function to defined them in a constant,
and add this constant in the function registry, rather than define the
function directly in the registry.

This allow for re-using the functions code and call it from other
functions.

## [REF] spreadsheet: refactor insert pivot methods

The goal here is to simplify the code of the insertion of a pivot in the
pivot core plugin. Now the domain/style for each cell of the pivot is computed
in the `SpreadsheetPivotTable` class, which simplify a lot the plugin.

## [IMP] speadsheet: add Pivot table function

Add an ODOO.PIVOT.TABLE function that is an array formula that returns the
whole pivot. This is needed for example to compute an average in a dashboard,
because standard pivot formula won't update with new values.

Task: [3318865](https://www.odoo.com/web#id=3318865&cids=1&menu_id=4720&action=333&active_id=2328&model=project.task&view_type=form)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/spreadsheet/static/src/pivot/pivot_table.js [hunk 1]
```diff
@@ -171,8 +180,91 @@ export class SpreadsheetPivotTable {
...
+        } else if (col === 0) {
+            const rowIndex = row - colHeadersHeight;
+            const domain = this._getRowDomain(rowIndex);
+            const indent = this._rows[rowIndex].indent;
+            //TODO DISCUSS: isn't undefined style on header a ugly ?
+            const style = indent <= 1 ? TOP_LEVEL_STYLE : indent === 2 ? HEADER_STYLE : undefined;
+            return { domain, isHeader: true, style };
+        } else {
+            const rowIndex = row - colHeadersHeight;
+            if (!includeTotal && this._isTotalRow(rowIndex)) {
```
Comment (hokolomopo) [inline addons/spreadsheet/static/src/pivot/pivot_table.js:228]: I've 2 different indexing for the rows (index in rows and index in whole pivot (headers + row)). I've no good ieads for the name. `rowIndex/fullPivotRowIndex` ? idk

File: addons/spreadsheet/static/src/pivot/pivot_table.js [hunk 2]
```diff
@@ -93,22 +102,22 @@ export class SpreadsheetPivotTable {
...
+        return this._cols[this.getColHeadersHeight() - 1].length;
     }
 
     /**
-     * Get the number of row in each columns
+     * Get the number of row in each column header
      * @return {number}
      */
-    getColHeight() {
+    getColHeadersHeight() {
```
Comment (anhe-odoo) [inline addons/spreadsheet/static/src/pivot/pivot_table.js:120]: Why did you change to `HeadersHeight` ? With this name, it looks like we are getting the height of the header part, while we are getting the number of rows, which is the height of the full column in term of cell

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 1 above)
Comment (anhe-odoo) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:228]: I would go for `rowIndex/fullRowIndex` or `pivotRowIndex/fullPivotRowIndex`

**Review by anhe-odoo — COMMENTED** [review]
LGTM :) Just a small question

File: addons/spreadsheet/static/src/pivot/pivot_functions.js [hunk 3]
```diff
@@ -7,6 +7,10 @@ import * as spreadsheet from "@odoo/o-spreadsheet";
 const { arg, toString } = spreadsheet.helpers;
 const { functionRegistry } = spreadsheet.registries;
 
+/**
+ * @typedef {import("@spreadsheet/pivot/pivot_table.js").SpreadsheetPivotTable} SpreadsheetPivotTable
+ */
+
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply addons/spreadsheet/static/src/pivot/pivot_functions.js:13]: Import the JS typing `SpreadsheetPivotTable` from `@spreadsheet/pivot/pivot_table.js`

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 2 above)
Comment (hokolomopo) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:120]: Because it _does_ return the height of the header part 🤔 `_cols.length` is the number of rows in the header part of the table

Comment (anhe-odoo) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:120]: Nevermind, when asking yesterday I was think of the header of a spreadsheet, not the header of a pivot, so I didn't undersand why a header could have a different number of row :D

File: addons/spreadsheet/static/src/pivot/pivot_table.js [hunk 4]
```diff
@@ -171,8 +180,80 @@ export class SpreadsheetPivotTable {
...
+            return { content: this._rowTitle, isHeader: true, style: HEADER_STYLE };
+        } else if (row <= colHeadersHeight - 1) {
+            const domain = this._getColHeaderDomain(col, row);
+            const style = row === colHeadersHeight - 1 ? MEASURE_STYLE : TOP_LEVEL_STYLE;
+            return { domain, isHeader: true, style };
+        } else if (col === 0) {
+            const rowIndex = row - colHeadersHeight;
+            const domain = this._getRowDomain(rowIndex);
+            const indent = this._rows[rowIndex].indent;
+            //TODO DISCUSS: isn't undefined style on header a ugly ?
```
Comment (rrahir) [inline addons/spreadsheet/static/src/pivot/pivot_table.js:223]: Personally, I don't really like it but turns out it will end up updating history with the value undefined whether you provide { style: undefined} or don't provide the key at all...

File: addons/spreadsheet/static/src/pivot/plugins/pivot_core_plugin.js [hunk 5]
```diff
@@ -273,76 +274,17 @@ export default class PivotCorePlugin extends CorePlugin {
      */
     _insertPivot(sheetId, anchor, id, table) {
```
Comment (rrahir) [inline addons/spreadsheet/static/src/pivot/plugins/pivot_core_plugin.js:275]: could we take this occasion to convert anchor to a `Position: {col, row}` Object ?

File: addons/spreadsheet/static/tests/pivots/model/pivot_plugin_test.js [hunk 6]
```diff
@@ -647,6 +648,65 @@ QUnit.module("spreadsheet > pivot plugin", {}, () => {
...
+                    <field name="product_id" type="col"/>
+                    <field name="foo" type="row"/>
+                    <field name="id" type="row"/>
+                    <field name="probability" type="measure"/>
+                </pivot>`,
+        });
+        const sheetId = model.getters.getActiveSheetId();
+        model.dispatch("CREATE_SHEET", { sheetId: "42" });
+        setCellContent(model, "A1", `=ODOO.PIVOT.TABLE("1")`, "42");
+        for (let col = 0; col <= 3; col++) {
```
Comment (rrahir) [inline addons/spreadsheet/static/tests/pivots/model/pivot_plugin_test.js:664]: replace 3 and 10 with the dimensions of the table, seems quite arbitrary. Also might be worth making sure that we don't spread beyond the range. I think you introduced `checkFunctionDoesntSpreadBeyondRange` in o-spreadsheet to that end?

File: addons/spreadsheet/static/src/pivot/pivot_table.js [hunk 7]
```diff
@@ -171,8 +180,91 @@ export class SpreadsheetPivotTable {
         return this._rows[index];
     }
 
-    getRowTitle() {
-        return this._rowTitle;
+    /**
+     * @returns {PivotCell[][]}
+     */
+    getPivotCells(includeTotal = true) {
```
Comment (rrahir) [inline addons/spreadsheet/static/src/pivot/pivot_table.js:186]: could we consider making this lazy? Each re-evaluation would go through this maybe without good reason?

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 1 above)
Comment (rrahir) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:228]: not sure but could you do with another offset?

**Review by rrahir — CHANGES_REQUESTED** [review]

File: addons/spreadsheet/static/src/pivot/pivot_functions.js [hunk 8]
```diff
@@ -37,110 +41,160 @@ function assertDomainLength(domain) {
...
-        compute: function () {
-            throw new Error(_t(`[[FUNCTION_NAME]] cannot be called from the spreadsheet.`));
-        },
-        returns: ["STRING"],
-        hidden: true,
-    });
+            const domain =
+                mode === "format" ? pivotCell.domain.map((d) => ({ value: d })) : pivotCell.domain;
+            const measure = mode === "format" ? { value: pivotCell.measure } : pivotCell.measure;
+            const fn = mode === "format" ? "computeFormat" : "compute";
```
Comment (rrahir) [inline addons/spreadsheet/static/src/pivot/pivot_functions.js:167]: outside of the loop

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 2 above)
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:120]: What about `getNumberOfHeaderRows` ?

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 4 above)
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:223]: It doesn't choque me plus que ça

File: addons/spreadsheet/static/src/pivot/plugins/pivot_core_plugin.js [hunk 9]
```diff
@@ -477,15 +359,21 @@ export default class PivotCorePlugin extends CorePlugin {
...
      * @param {number} col
      * @param {number} row
-     * @param {string} formula
-     * @param {Array<string>} args
+     * @param {string} pivotId
+     * @param {PivotCell} pivotCell
      */
-    _addPivotFormula(sheetId, col, row, formula, args) {
+    _addPivotFormula(sheetId, pivotId, col, row, pivotCell) {
+        const formula = pivotCell.isHeader ? "ODOO.PIVOT.HEADER" : "ODOO.PIVOT";
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/pivot/plugins/pivot_core_plugin.js:366]: `formula` => `function`

File: addons/spreadsheet/static/src/pivot/pivot_functions.js [hunk 10]
```diff
@@ -147,8 +151,50 @@ const ODOO_PIVOT_POSITION = {
...
+    description: _t("Get a pivot table."),
+    args: [
+        arg("pivot_id (string)", _t("ID of the pivot.")),
+        arg(
+            "include_total (boolean, default=TRUE)",
+            _t("Whether to include total/sub-totals or not.")
+        ),
+    ],
+    compute: function (pivotId, includeTotal) {
+        const _includeTotal = includeTotal ?? true;
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/pivot/pivot_functions.js:184]: `toBoolean`

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 7 above)
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:186]: Nice catch
or just compute it in the constructor and store it (readonly property) ? It's ok to store a computed/derived value if the data is immutable/all read only

File: addons/spreadsheet/static/tests/pivots/model/pivot_plugin_test.js [hunk 11]
```diff
@@ -647,6 +648,65 @@ QUnit.module("spreadsheet > pivot plugin", {}, () => {
...
+                assert.equal(cellPivotTableFormula.value, cellStandardInsertPivot.value);
+                assert.equal(
+                    cellPivotTableFormula.format || "",
+                    cellStandardInsertPivot.format || ""
+                );
+            }
+        }
+    });
+
+    QUnit.test("PIVOT.TABLE results with no_total argument", async function (assert) {
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/tests/pivots/model/pivot_plugin_test.js:678]: `no_total` => `include_total=FALSE`

**Review by LucasLefevre — COMMENTED** [review]
Nice work :+1: 
The REF is good (and even negative diff)

File: addons/spreadsheet/static/src/pivot/pivot_table.js [hunk 12]
```diff
@@ -171,8 +180,80 @@ export class SpreadsheetPivotTable {
         return this._rows[index];
     }
 
-    getRowTitle() {
-        return this._rowTitle;
+    /**
+     * @returns {PivotCell[][]}
+     */
+    getPivotCells() {
+        const pivotHeight = this.getColHeadersHeight() + this.getRowHeight();
```
[the remark this answers is not in the document]
Comment (rrahir) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:187]: I like getNumberOfDataRows

File: addons/spreadsheet/static/src/pivot/pivot_table.js (hunk 4 above)
Comment (rrahir) [reply addons/spreadsheet/static/src/pivot/pivot_table.js:223]: let's keep it as it is then

**Comment by LucasLefevre** [comment]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-ff
